### PR TITLE
Stepper motor

### PR DESCRIPTION
**Pull request by zignig**
Stepper motor, 

TODO
a collection to pull out NEMA## sizes.
need a cutout for mounting
even 4 fasteners so it auto mounts to whatever it is parented to.

### CODE REVIEW

File: src/cqparts_motors/stepper.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+import cqparts
+from cqparts.params import *
+from cqparts.display import render_props, display
+from cqparts.constraint import Fixed, Coincident
+from cqparts.constraint import Mate
+from cqparts.utils.geometry import CoordSystem
+
+
+class Axle(cqparts.Part):
```
Comment (fragmuffin) [inline src/cqparts_motors/stepper.py]: suggested change to `_Axle`

File: src/cqparts_motors/stepper.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+            .chamfer(0.4)
+        return ax
+
+    def get_cutout(self, clearance=0):
+        return cq.Workplane('XY', origin=(0, 0, 0)) \
+            .circle((self.diam / 2) + clearance) \
+            .extrude(7)
+
+
+class EndCap(cqparts.Part):
```
Comment (fragmuffin) [inline src/cqparts_motors/stepper.py]: suggested change to `_EndCap`

File: src/cqparts_motors/stepper.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    @property
+    def mate_bottom(self):
+        return Mate(self, CoordSystem(
+            origin=(0, 0, -self.height/2),
+            xDir=(0, 1, 0),
+            normal=(0, 0, 1)
+            ))
+
+
+class Stator(cqparts.Part):
```
Comment (fragmuffin) [inline src/cqparts_motors/stepper.py]: suggested change to `_Stator`

File: src/cqparts_motors/stepper.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    @property
+    def mate_bottom(self):
+        return Mate(self, CoordSystem(
+            origin=(0, 0, -self.height/2),
+            xDir=(0, 1, 0),
+            normal=(0, 0, 1)
+            ))
+
+
+class StepperMount(EndCap):
```
Comment (fragmuffin) [inline src/cqparts_motors/stepper.py]: suggested change to `_StepperMount`

File: src/cqparts_motors/stepper.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        obj.faces(">Z").workplane() \
+            .rect(self.spacing, self.spacing, forConstruction=True)\
+            .vertices() \
+            .hole(self.hole_size)
+        obj.faces(">Z").workplane()\
+            .circle(self.boss/2).extrude(2)
+        return obj
+
+
+class Back(EndCap):
```
Comment (fragmuffin) [inline src/cqparts_motors/stepper.py]: suggested change to `_Back`

File: src/cqparts_motors/stepper.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
+
+
+import cadquery as cq
+
+import cqparts
+from cqparts.params import *
+from cqparts.display import render_props, display
```
Comment (fragmuffin) [inline src/cqparts_motors/stepper.py]: no need to import `display`

File: src/cqparts_motors/stepper.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        axle = self.components['axle']
+        top = self.components['topcap']
+        local_obj = top.local_obj
+        local_obj = local_obj.cut(axle.get_cutout(clearance=0.5))
+
+    def make_alterations(self):
+        self.apply_cutout()
+
+s = Stepper()
+display(s)
```
Comment (fragmuffin) [inline src/cqparts_motors/stepper.py]: Shouldn't create an instance, or display `Stepper` in a python module

**Comment by fragmuffin** [comment]
@zignig A few changes there.

**Sphinx documentation**
All those classes will be automatically published with `sphinx` [somewhere here](https://fragmuffin.github.io/cqparts/doc/api/cqparts_motors.html)
The above suggestions to prefix the classes with a `_` is to remove them from being documented.

**Fasteners**
neat idea for the 4 fasteners ;), make them optional though so others can use it how they want.

File: src/cqparts_motors/stepper.py (hunk 1 above)
Comment (zignig) [reply src/cqparts_motors/stepper.py]: I think this should be moved out as a Shaft, there are lots of different shaft set ups that we can use. And having mattes provided by the shaft would be helpful with larger constructs.

**Comment by zignig** [comment]
Okay fixes are in, have some dc motor work for the next one.

**Comment by fragmuffin** [comment]
Sorry for the delay @zignig 
I've been sick and working hard... a combination that doesn't leave much room for this.
I'll review tonight (about 5hrs from now)

**Comment by zignig** [comment]
no worries :)

I'm in Perth, so a few hours behind.

**Comment by fragmuffin** [comment]
I made a few changes, I hope you don't mind.

### Catalogue
You added some standard sizes as function-calls within the `cqparts_motors.stepper` module in 4f7a0cc44ec2d30dc7355a2a173eb7b03b50e57a.
I've moved them out of the stepper's "blue-print" module into a [`JSONCatlogue`](https://fragmuffin.github.io/cqparts/doc/api/cqparts.catalogue.html#cqparts.catalogue.JSONCatalogue).

```python
>>> from cqparts.catalogue import JSONCatalogue
>>> import cqparts_motors
>>> filename = os.path.join(
...     os.path.dirname(cqparts_motors.__file__),
...     'catalogue', 'stepper-nema.json',
... )
>>> catalogue = JSONCatalogue(filename)
>>> item = catlaogue.get_query()
>>> stepper = catalogue.get(item.criteria.size == 17)
>>> from cqparts.display import display
>>> display(stepper)
```

Note: there are many things that can be improved about the way catalogues are used (ie: the above code is messy), but that's independent of how or why they're made.

### Testing
The added advantage of using a catalogue means we can automatically test each catalogue item is build with some semblance of sanity with [`CatalogueTest`](https://fragmuffin.github.io/cqparts/doc/api/cqparts.utils.html#cqparts.utils.test.CatalogueTest)

Tests can be run locally with:

```bash
$ cd tests
$ python runtests.py --pattern test_motors*
test_NEMA_Stepper_11 (t_cqparts_motors.test_motors_catalogue.CatalogueTest_stepper-nema) ... ok
test_NEMA_Stepper_14 (t_cqparts_motors.test_motors_catalogue.CatalogueTest_stepper-nema) ... ok
test_NEMA_Stepper_17 (t_cqparts_motors.test_motors_catalogue.CatalogueTest_stepper-nema) ... ok
test_NEMA_Stepper_23 (t_cqparts_motors.test_motors_catalogue.CatalogueTest_stepper-nema) ... ok
test_NEMA_Stepper_8 (t_cqparts_motors.test_motors_catalogue.CatalogueTest_stepper-nema) ... ok

----------------------------------------------------------------------
Ran 5 tests in 0.825s

OK
```

If you're happy with that, I'd be happy to merge it in.

**Comment by zignig** [comment]
Looks great , Splitting the catalogue out make much more sense if the testing system can iterate through them. 

I'm happy for you to merge.

I will copy your lead on on this and fold it into the  DC motor work I have done so far and make it catalogue(able).

**Comment by fragmuffin** [comment]
@zignig thank you so much for your contribution!
Congrats for being the first contributor of new components to `cqparts` :trophy:

**Pull request merged**